### ISSUE TITLE
Support exporting parsers as ES2015 modules

### ISSUE THREAD

**Issue by fatfisz** [pegjs/pegjs#423]
Currently the `buildParser` method returns either an object or an IIFE. This prevents seamless ES6-style imports.

There is a plugin that allows importing parsers built from grammar files, `rollup-plugin-pegjs`. It currently can't do much more than either `export default generatedParser;` or `module.exports = generatedParser;`.

Despite the claims in the `rollup-plugin-pegjs` README that

``` js
import { parse } from "grammar.pegjs";
```

is possible, I've been unable to do so and have to rely on this:

``` js
import parser from "grammar.pegjs";
const { parse } = parser;
```

So taking that into consideration, would it be ok if there was another option for `output`: `source-es6` or `source-es6-module` that would have IIFE stripped and would have ES6-style exports instead of returning an object?

I could do the PR if the idea is accepted.

**Comment by dmajda** [comment]
The way I see it, there are two questions in play here:

  1. What is the responsibility of the `buildParser` function?
  2. What exactly should be exported?

As for 1, in the current architecture `buildParser` returns either the parser object or source code which produces it when evaluated. This invariant is [encoded into the compiler](https://github.com/pegjs/pegjs/blob/0c39f1cf86d73c85a1f832e6cfddf4176f9e58bc/lib/compiler.js#L59..L62). The function doesn’t deal with the module system at all (this is [done  in `bin/pegjs`](https://github.com/pegjs/pegjs/blob/0c39f1cf86d73c85a1f832e6cfddf4176f9e58bc/bin/pegjs#L270)).

This design is simple and understandable, but your request made me realize that it’s probably not the most practical one. The code which handles exporting currently has to be replicated in many tools which use PEG.js API (e.g. [`gulp-peg`](https://github.com/lazutkin/gulp-peg/blob/d76f75549ca88e3e1f58b0418dc55c6acdf5a374/src/index.coffee#L10) or [`grunt-peg`](https://github.com/dvberkel/grunt-peg/blob/7438cc4cba68fe3fd256a8076019d7f3876e2416/tasks/peg.js#L39)). It would be more useful to put export handling into `buildParser` itself (either by allowing more `output` values or by adding specialized option(s) valid when `output` is set to `"source"`). I’ll have a look at this as part of #362 (I just added an item into a checklist there).

As for 2, the core issue is whether generated parsers are *objects* or *modules*. In the pre-ES6 Node.js world this question doesn’t have to be answered because modules are represented by objects and export works the same for both. In the ES6 world the situation is different. If generated parsers are objects, the proper way to export them is using the default export. Import and usage would then look something like this:

```js
import parser from "parser";

try {
  parser.parse(...);
} catch (e) {
  if (e instanceof parser.SyntaxError) {
    // ...
  } else {
    // ...
  }
}
```

On the other hand, if generated parsers are modules, the proper way to export them is using named exports. Import and usage would then look something like this:

```js
import { parse, SyntaxError } from "parser";

try {
  parse(...);
} catch (e) {
  if (e instanceof SyntaxError) {
    // ...
  } else {
    // ...
  }
}
```

or, alternatively:

```js
import * as parser from "parser";

try {
  parser.parse(...);
} catch (e) {
  if (e instanceof parser.SyntaxError) {
    // ...
  } else {
    // ...
  }
}
```

The current code assumes that generated parsers are objects — you can even reference the parser object from a grammar as `parser`. I’d like to keep things that way, at least for the time being.

In practice, I think it means adding ES6-style exports to generated parsers is not necessary at this point. The only practical way to use ES6 modules right now is through transpilers, and at least Babel (with the `es2015` preset) seems to be able to correctly handle both modules that use `module.exports = ...` (the current way) and those that use `export default ...` (the ES6 way). So I don’t think any change is needed here, at least form the practical standpoint.

I’ll leave this issue open in case the problem is actually more complex than I think. Also, I assume adding support for ES6 modules will become necessary at some later point when they will be supported natively more widely, so the issue will need to be resolved eventually.

**Comment by fatfisz** [comment]
I agree that this is not necessary and it is possible to use `pegjs` as it is now, even with `rollup`.

The main benefit of ES6-style exports is removing the unnecessary IIFE. Also `rollup` removes any exports that are not used anywhere, but I'm not so experienced with `pegjs` as to know whether this is a big optimalization. Using an IIFE makes static analysis much harder, if not impossible.

Anyway, I've since forked this repo and applied the changes I needed (or more exactly: my friend needed), so you can see the results here for future reference if you decide it's ok to go the ES6-style exports way:
* [The main commit](https://github.com/fatfisz/pegjs/commit/fd798f1541eca052c4bd546371b1c2d6b18b06e7?w=1)
* [The list of commits (I messed one thing up, so 2 commits in total)](https://github.com/fatfisz/pegjs/commits/es6-support)

---
I also had an idea of how to (maybe) improve the current architecture: the `generate-js` pass could return only the code containing functions, objects, etc.; exporting could be handled in another pass. This would improve readability a bit, since the files are huge now. I also think it would be less painful to deal with exports when they would be separated from the parser-generating logic.

**Comment by jonaskello** [comment]
I would like to have a `--format esm` option that outputs a ECMA Script Module format (aka as es6 module format) compatible file. I need this to import it seamlessly in typescript.

I think this should be quite simple to implement. Looking at the commonjs output it does:

```js
  return {
    SyntaxError: peg$SyntaxError,
    parse:       peg$parse
  };
```

So instead just remove the IIFE and do:

```js
    export const SyntaxError = peg$SyntaxError,
    export const parse = peg$parse
```

Should we do a PR for this?

**Comment by dmajda** [comment]
It seems like the demand for ES2015-style parser modules is bigger than I thought.

In a [previous comment](https://github.com/pegjs/pegjs/issues/423#issuecomment-214762654), I discussed whether the parser really is an object or a module and what exactly should be exported. I concluded that:

> The current code assumes that generated parsers are objects — you can even reference the parser object from a grammar as `parser`. I’d like to keep things that way, at least for the time being.

I think it’s time to reverse this position. Generated parsers should be treated as modules exporting the `parse` function and the `SyntaxError` constructor (and optionally the `DefaultTracer` constructor). Note the `parser` variable mentioned in the comment is not available anymore (since 0.10.0).

@jonaskello You or anyone else are welcome to submit a PR implementing this. It should be trivial.

**Comment by fatfisz** [comment]
Since I've already created a fork with the required changes some time ago, I'd happily update it and prepare the PR. The problem is, I can do this at earliest a week from now.

If anyone wants to take care of this in the meanwhile, I've left some reference links to my changes in the previous comments.

**Comment by jonaskello** [comment]
I have no urgent need for this at the moment. @fatfisz If you already have the changes needed I think it would be best if you do it.

**Comment by StoneCypher** [comment]
I propose that this be re-opened.  This has fairly significant consequences for tree shaking.

**Comment by StoneCypher** [comment]
@futagoza - please consider reopening this.

**Comment by futagoza** [comment]
@StoneCypher Why reopen this? It's already been implemented in v0.11, I just need to find more time to finish of the test suite re-write before I release v0.11

**Comment by StoneCypher** [comment]
I didn't know it was already implemented.  

I apologize for the noise.

**Comment by StoneCypher** [comment]
So, I'm kind of annoyed, because you told me to not re-open this in May of 2019 because it was already done, but in February of 2020, it's actually not on NPM yet, and you just said a month ago that version 0.11.0 is never happening

This needs to be re-opened, and left open until it's actually done.  Done-in-a-failed-branch doesn't count

Cleaning up this repo's issue tracker to re-open all the things that shouldn't be closed is going to be terrible

**Comment by StoneCypher** [comment]
The correct fix for this is easily ported into `0.10.0` and published.  It's a six line copy paste, and it's very high stakes for a lot of us.

Why haven't you done that yet?  We shouldn't be waiting on your `0.11.0`, your `0.12.0`, your `OpenPeg`, your `PegX`, et cetera

Please re-open this issue, and add a maintainer who can make this project healthy again.

**Comment by leifg** [comment]
What's still missing to release this?

I'm willing to help out!

**Comment by StoneCypher** [comment]
The new maintainer has never released since he took over, and ghosted months ago
